File: Pa1/Lex.java
// Lex.java
// Diego Magdaleno
// dmmagdal
// pa1: Takes file input and sorts the data lexicographically (in order) using the List ADT

import java.util.Scanner;
import java.io.*;

public class Lex{
	public static void main(String[] args) throws IOException, FileNotFoundException, NullPointerException{

		// if there aren't exactly two arguments, print this error message and exit the program
		if (args.length != 2){
			System.err.println("Usage: Lex infile outfile");
			System.exit(1);
		}

		// otherwise, continue on

		Scanner in = new Scanner(new File(args[0]));								// variable to take input from the input file 
		PrintWriter out = new PrintWriter(new FileWriter(args[1]));					// variable to write to the output file
		BufferedReader reader = new BufferedReader(new FileReader(args[0]));		// variable to read through the input file

		
		int lineCount = 0;															// int to hold number of lines in the input file
		// while the reader and can read through a line, increment the line count
		while (reader.readLine() != null){
			lineCount++;
		}

		// String array to hold data from input file
		String[] input = new String[lineCount];										// string array to hold all input from the input file, uses lineCount variable from above for the size 
		int index = 0;																// int holds the index for thie input[] string array, starting from zero
		// while the input variable has a next line, store the line to the input[] string array and increment the index variable
		while (in.hasNextLine()){
			input[index] = in.nextLine();
			index++;
		}

		/* // test to see if the input was being copied correctly to the string array. Gives coorect output
		for (int j = 0; j < input.length; j++){
			out.println(input[j]);
			System.out.println(input[j]);

		}*/

		List indices = new List();													// initialize new list object to store indices
		// start the sorting
		indices.append(0);	// input[0]												insert the first item
		
		for (int i = 1; i < input.length; i++){										// iterate through the inut String[] array
			indices.moveFront();													// put the cursor to the front
			if (input[i].compareTo(input[indices.get()]) < 0){						// if the item is less than the front, insert the index before the cursor
				indices.insertBefore(i);
			}
			else {	//arg[i].compareT0(arg[l.get()]) > 0							// here the items in the array are greater than the cursor
				while (input[i].compareTo(input[indices.get()]) > 0){				// while the above is true
					if (indices.get() == indices.back()){							// if the cursor is at the back node, insert the index after
						indices.insertAfter(i);
					}
					else{															// otherwise, increment the cursor to the next node on the list
						indices.moveNext();
					}
				}
				indices.insertBefore(i);											// by now, you will insert the index before the cursor
			}
		}																			// end sort

		// print to file
		indices.moveFront();														// set cursor to front
		/* // tried to iterate through the list to print. ended up having copy of the last node at the end. delete this chunk
		for (int l = 0; l < indices.length(); l++){									// iterate through list and return the umber of the index to be printed
			out.println(input[indices.get()]);
			// System.out.println(input[indices.get()]);							// print to console to see what gets printed to file. remember to comment out in final product
			indices.moveNext();														// move the cursor to the next position in the list
		}
		*/
		while (indices.index() != indices.length()-1){								// while the index of the cursor is not at the end of the list
			out.println(input[indices.get()]);										// print to file the index held in the list of the input string[] array
			indices.moveNext();														// move the cursor to the next position in the list
		}

		
		// close the variables to read, scan, and write the files
		in.close();
		out.close();
		reader.close();
	}
}

File: Pa2/backup/Lex.c
// Lex.c
// Diego Magdaleno
// dmmagdal
// pa2: Sorts strings in a file using list ADT

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "List.h"

#define MAX_LEN 100

int main(int argc, char *argv[])
{
	// check command line for correct number of arguments
	if (argc != 3){
		fprintf(stderr, "Usage: %s <input file> <output file> \n", argv[0]);
		exit(EXIT_FAILURE);
	}


	int lineCount = 0;								// hold linecount variables
	FILE *in, *out;									// pointers to input and output files
	List indexList = newList();						// create new list to store the indices
	char line[MAX_LEN];								// hold the string of each line
	char* lineinput[MAX_LEN];						// holds all the input from each line of the file


	// open files for reading and writing
	in = fopen(argv[1], "r");						// open input file to read
	out = fopen(argv[2], "w");						// open output file to write to
	if (in == NULL){
		printf("Unable to open file %s for reading\n", argv[1]);
		exit(EXIT_FAILURE);
	}
	if (out == NULL){
		printf("Unable to open file %s for writing\n", argv[2]);
		exit(EXIT_FAILURE);
	}


	// count the lines and store each line to the array
	while (fgets(line, MAX_LEN, in) != NULL){
		lineinput[lineCount] = strdup(strtok(line, "\n"));
		// printf("%s\n", lineinput[lineCount]);	// test storage of line to array
		lineCount++;
	}


	// copy array to smaller sized array
	char* input[lineCount];
	for (int i = 0; i < lineCount; i++){
		input[i] = lineinput[i];
		// printf("%s\n", input[i]);				// test if array copied correctly
	}

	// sorting algorithm
	append(indexList, 0);
	for (int i = 1; i < lineCount; i++){
		moveFront(indexList);
		if (strcmp(input[i], input[get(indexList)]) < 0){
			insertBefore(indexList, i);
		}
		else {
			while (strcmp(input[i], input[get(indexList)]) > 0){
				if (get(indexList) == back(indexList)){
					insertAfter(indexList, i);
				}
				else {
					moveNext(indexList);
				}
			}
			insertBefore(indexList, i);
		}
	}

	// print to file 
	moveFront(indexList);
	while (index(indexList) != length(indexList)-1){
		fprintf(out, "%s\n", input[get(indexList)]);
		// printf("%s\n", input[get(indexList)]);	// test if sort worked correctly
		moveNext(indexList);
	}



	// clean up te memory
	clear(indexList);
	freeList(&indexList);
	fclose(in);
	fclose(out);

	exit(EXIT_SUCCESS);
} 